### ISSUE TITLE
Pre-processing of projects on server

### ISSUE THREAD

**Issue by ricschuster** [NCC-CNC/wheretowork#181]
The comments from BC staff made we wonder whether it would be possible to have the app ingest/process the data that's stored on the server and thus allow it to load faster?

Not sure if I explain this well, but the idea is to have the Where to Work do its pre-processing steps on folders on the server, save the workspace and use that to speed up loading. 

@jeffreyhanson what do you think about that idea?

**Comment by jeffreyhanson** [comment]
Yeah, I like the idea of using the server to help with data processing. I think the best way to do this would probably involve creating a seperate data formatting app (rather than adding this functionality to the where to work app). In this data formatting app, the users could upload their data, specify which attributes/layers correspond to themes, features, includes, and then have it process the data on the server.  I worry that adding this functionality to the where to work tool directly would make it very complicated to use.

**Comment by jeffreyhanson** [comment]
Hmm, actually, now that I think about it - would we actually need such a data formatting app given the plugin that @JCLaurence has been working on? Since the plugin will provide users with the ability to format project data using an interactive interface - won't that provide the same functionality? I guess one difference is that  the plugin would be running on the users computer locally - maybe there's someway to create dedicated Windows server to host the plugin?

**Comment by ricschuster** [comment]
Sorry, I didn't explain this well. You are right, we have the functionality you are thinking about with JC's tool.
What I meant was using the YAML, spatial data, and attribute data to let the Where to Work tool do the data ingestion so to speak.

All the processing (including building the leaflet map) that Where to Work does every time you load a project could be done once. At the end of the process Where to Work would save an RData file (or something similar) to the project folder on the server. Next time the project gets loaded, Where to Work would not have to go through the entire process, but just load the RData file.

Does that make sense?

**Comment by jeffreyhanson** [comment]
Oh I see! Hmm, that might be possible? I don't know exactly what the main bottleneck is for larger datasets - so I don't know if caching the data import process will help much? E.g. I know this would be very useful for large shapefiles, but we've seen slow performance with raster data too? So think it would probably be worth running some benchmarks first to work out where the main bottleneck is. Also, if the plan is to ultimately replace leaflet, then maybe it's worth waiting until that's finsihed before implementing this?

**Comment by ricschuster** [comment]
I like the idea of benchmarking. I will try to look into this, so you can focus on other things. 
Good point about waiting for a switch away from leaflet. Will think about that. I am a bit worried that this might happen anytime soon though, given what you mentioned about potential timelines.
